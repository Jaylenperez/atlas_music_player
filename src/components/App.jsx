import Footer from "./Footer";
import MusicPlayer from "./MusicPlayer";

function App() {
  return (
    <div className="flex h-full min-h-screen flex-col justify-between bg-amber-100 p-8 dark:bg-purple-950">
      <MusicPlayer />
      <Footer />
    </div>
  );
}

export default App;
