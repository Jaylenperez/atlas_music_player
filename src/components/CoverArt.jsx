import placeholder from "../assets/placeholder.svg";

export default function CoverArt() {
  return (
    <figure className="aspect-square w-full overflow-hidden rounded-md bg-neutral-200 shadow ring-1 ring-black/10 dark:bg-neutral-800 dark:ring-white/10">
      <img
        src={placeholder}
        alt="Album cover art"
        className="h-full w-full object-cover"
        loading="lazy"
      />
    </figure>
  );
}
