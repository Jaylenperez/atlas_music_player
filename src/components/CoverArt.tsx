import React from "react";

export interface CoverArtProps {
  src: string;
  alt?: string;
}

export const CoverArt: React.FC<CoverArtProps> = ({ src, alt = "Cover Art" }) => {
  return (
    <div className="flex justify-center md:justify-start">
      <img
        src={src}
        alt={alt}
        className="aspect-square w-full rounded-md object-cover sm:w-72 md:w-80 lg:w-96 xl:w-[50rem]"
      />
    </div>
  );
};
