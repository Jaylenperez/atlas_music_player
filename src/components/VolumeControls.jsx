import { Volume2 } from "lucide-react";

export default function VolumeControls() {
  return (
    <div className="w-full pl-1">
      <div className="flex items-center gap-3">
        {/* Volume Button */}
        <button
          type="button"
          aria-label="Volume"
          title="Volume"
          className="p-2 text-purple-900 dark:text-amber-200"
        >
          <Volume2 className="h-5 w-5" />
        </button>

        <input type="range" className="w-full" />
      </div>
    </div>
  );
}
