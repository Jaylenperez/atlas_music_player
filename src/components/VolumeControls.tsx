import React from "react";

export interface VolumeControlsProps {
  lightMode: boolean;
  volume: number;
  onVolumeChange: (volume: number) => void;
}

export const VolumeControls: React.FC<VolumeControlsProps> = ({
  lightMode,
  volume,
  onVolumeChange,
}) => {
  const iconBtn =
    "rounded p-2 transition " +
    (lightMode
      ? "text-gray-800 hover:bg-gray-200"
      : "text-amber-300 hover:bg-indigo-900");

  return (
    <div className="flex w-full items-center gap-2">
      <button className={iconBtn} aria-label="Volume">
        <svg
          xmlns="http://www.w3.org/2000/svg"
          className="h-5 w-5"
          fill="none"
          viewBox="0 0 24 24"
          stroke="currentColor"
        >
          <path d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z" />
          <path d="M16 9a5 5 0 0 1 0 6" />
          <path d="M19.364 18.364a9 9 0 0 0 0-12.728" />
        </svg>
      </button>
      <input
        type="range"
        min={0}
        max={100}
        value={volume}
        onChange={(e) => onVolumeChange(Number(e.target.value))}
        aria-label="Volume slider"
        className={`h-1 w-full cursor-pointer ${
          lightMode ? "accent-indigo-800" : "accent-amber-200"
        }`}
      />
    </div>
  );
};
