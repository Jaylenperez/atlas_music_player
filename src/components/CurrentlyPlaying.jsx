import CoverArt from "./CoverArt";
import SongTitle from "./SongTitle";
import PlayControls from "./PlayControls";
import VolumeControls from "./VolumeControls";

export default function CurrentlyPlaying() {
  return (
    <div className="flex flex-col items-center gap-4">
      <CoverArt />

      <div className="w-full max-w-[400px] text-left">
        <SongTitle />
      </div>

      <div className="text-plum">
        <PlayControls />
      </div>

      <VolumeControls />
    </div>
  );
}
