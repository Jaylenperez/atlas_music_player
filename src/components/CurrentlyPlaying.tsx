import React, { useEffect, useRef, useState } from "react";
import { CoverArt } from "./CoverArt";
import { SongTitle } from "./SongTitle";
import PlayControls from "./PlayControls";
import { VolumeControls } from "./VolumeControls";
import { PlaylistSong } from "./App";

interface Song extends PlaylistSong {
  cover: string;
  song: string;
}

interface Props {
  lightMode: boolean;
  track: PlaylistSong;
  playlist: PlaylistSong[];
  onSelect: (t: PlaylistSong) => void;
}

const CurrentlyPlaying: React.FC<Props> = ({
  lightMode,
  track,
  playlist,
  onSelect,
}) => {
  const [fullSong, setFullSong] = useState<Song | null>(null);
  const [shuffleEnabled, setShuffleEnabled] = useState(false);
  const [volume, setVolume] = useState(50);
  const [isPlaying, setIsPlaying] = useState(false);
  const [playbackSpeed, setPlaybackSpeed] = useState<0.5 | 1 | 2>(1);

  const audioRef = useRef<HTMLAudioElement | null>(null);

  // Load song details
  useEffect(() => {
    fetch(`/api/v1/songs/${track.id}`)
      .then((r) => {
        if (!r.ok) throw new Error("Song fetch failed");
        return r.json();
      })
      .then((data: Song) => setFullSong(data))
      .catch(() => setFullSong(null));
  }, [track.id]);

  // When fullSong changes, swap audio
  useEffect(() => {
    if (!fullSong) return;
    audioRef.current?.pause();
    const audio = new Audio(fullSong.song);
    audio.volume = volume / 100;
    audio.playbackRate = playbackSpeed;   // set initial speed
    audioRef.current = audio;
    audio.play().then(() => setIsPlaying(true)).catch(() => {});
    return () => {
      audio.pause();
    };
  }, [fullSong]);

  // Update volume live
  useEffect(() => {
    if (audioRef.current) {
      audioRef.current.volume = volume / 100;
    }
  }, [volume]);

  // Update playbackRate when speed changes
  useEffect(() => {
    if (audioRef.current) {
      audioRef.current.playbackRate = playbackSpeed;
    }
  }, [playbackSpeed]);

  // Playback handlers
  const togglePlay = () => {
    if (!audioRef.current) return;
    if (isPlaying) {
      audioRef.current.pause();
      setIsPlaying(false);
    } else {
      audioRef.current
        .play()
        .then(() => setIsPlaying(true))
        .catch(() => {});
    }
  };

  // Back / Forward / Shuffle logic
  const idx = playlist.findIndex((t) => t.id === track.id);
  const isFirst = idx <= 0;
  const isLast = idx === playlist.length - 1;

  const handleBack = () => {
    if (!isFirst) onSelect(playlist[idx - 1]);
  };
  const handleForward = () => {
    if (shuffleEnabled) {
      const rand = Math.floor(Math.random() * playlist.length);
      onSelect(playlist[rand]);
    } else if (!isLast) {
      onSelect(playlist[idx + 1]);
    }
  };
  const toggleShuffle = () => setShuffleEnabled((s) => !s);

  const coverSrc = fullSong?.cover ?? "/placeholder.svg";

  return (
    <div className="flex h-full w-full flex-col justify-between px-4 py-4">
      <div className="flex-shrink-0">
        <CoverArt src={coverSrc} />
      </div>

      {fullSong && (
        <div className="space-y-3">
          <SongTitle
            lightMode={lightMode}
            title={fullSong.title}
            artist={fullSong.artist}
          />
          <PlayControls
            lightMode={lightMode}
            isPlaying={isPlaying}
            onPlayToggle={togglePlay}
            onBack={handleBack}
            onForward={handleForward}
            disableBack={isFirst}
            disableForward={!shuffleEnabled && isLast}
            shuffleEnabled={shuffleEnabled}
            onShuffleToggle={toggleShuffle}
            playbackSpeed={playbackSpeed}
            onSpeedChange={setPlaybackSpeed}
          />
        </div>
      )}

      <div className="flex-shrink-0">
        <VolumeControls
          lightMode={lightMode}
          volume={volume}
          onVolumeChange={setVolume}
        />
      </div>
    </div>
  );
};

export default CurrentlyPlaying;
