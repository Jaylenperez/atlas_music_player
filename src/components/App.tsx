import React, { useState, useEffect } from "react";
import CurrentlyPlaying from "./CurrentlyPlaying";
import Playlist from "./Playlist";

export interface PlaylistSong {
  id: string;
  title: string;
  artist: string;
  genre: string;
  duration: number;
}

const App: React.FC = () => {
  const [lightMode, setLightMode] = useState<boolean>(false);
  const [playlist, setPlaylist] = useState<PlaylistSong[]>([]);
  const [selectedTrack, setSelectedTrack] = useState<PlaylistSong | null>(null);

  useEffect(() => {
    fetch("/api/v1/playlist")
      .then((res) => res.json())
      .then((data: PlaylistSong[]) => {
        setPlaylist(data);
        if (data.length > 0) setSelectedTrack(data[0]);
      })
      .catch((err) => console.error("Failed to load playlist:", err));
  }, []);

  return (
    <div
      className={`${
        lightMode ? "bg-amber-100 text-gray-900" : "bg-indigo-950 text-white"
      } flex h-screen flex-col overflow-x-hidden overflow-y-auto md:overflow-y-hidden md:flex-row`}
    >
      <button
        onClick={() => setLightMode((m) => !m)}
        className={`absolute top-4 right-4 rounded px-3 py-1 text-sm font-medium transition ${
          lightMode
            ? "bg-slate-400 text-gray-800"
            : "bg-gray-800 text-gray-200"
        }`}
      >
        {lightMode ? "Dark Mode" : "Light Mode"}
      </button>

      <div className="flex w-full flex-col justify-between p-4 md:w-1/2 md:p-8">
        {selectedTrack && (
          <CurrentlyPlaying
            lightMode={lightMode}
            track={selectedTrack}
            playlist={playlist}
            onSelect={setSelectedTrack}
          />
        )}
      </div>

      <div className="w-full p-4 md:w-1/2 md:border-t-0 md:p-8 md:overflow-y-auto">
        <Playlist
          lightMode={lightMode}
          tracks={playlist}
          selected={selectedTrack}
          onSelect={setSelectedTrack}
        />
      </div>
    </div>
  );
};

export default App;
