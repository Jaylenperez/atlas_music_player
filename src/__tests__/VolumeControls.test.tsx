// src/__tests__/VolumeControls.test.tsx
import React from "react";
import { render } from "@testing-library/react";
import { test, expect, vi } from "vitest";
import { VolumeControls } from "../components/VolumeControls"; // adjust path if needed

test("VolumeControls renders correctly", () => {
    // Create a mock function for onVolumeChange
    const mockOnVolumeChange = vi.fn();

    const { container } = render(
        <VolumeControls
            lightMode={true}
            volume={50}
            onVolumeChange={mockOnVolumeChange}
        />
    );

    // Snapshot test
    expect(container).toMatchSnapshot();
});
