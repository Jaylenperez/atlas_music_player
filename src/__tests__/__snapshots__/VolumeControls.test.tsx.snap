// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VolumeControls renders correctly 1`] = `
<div>
  <div
    class="flex w-full items-center gap-2"
  >
    <button
      aria-label="Volume"
      class="rounded p-2 transition text-gray-800 hover:bg-gray-200"
    >
      <svg
        class="h-5 w-5"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
        />
        <path
          d="M16 9a5 5 0 0 1 0 6"
        />
        <path
          d="M19.364 18.364a9 9 0 0 0 0-12.728"
        />
      </svg>
    </button>
    <input
      aria-label="Volume slider"
      class="h-1 w-full cursor-pointer accent-indigo-800"
      max="100"
      min="0"
      type="range"
      value="50"
    />
  </div>
</div>
`;
